### PR TITLE
JAMES-2937 RabbitMQ startvation: rely more on reactor primitives

### PR DESCRIPTION
**Pull request by chibenwa**
This is a real POC! Take it as is....

Not enough time to continue that work today!

 - I blew up the metric contract -> fix it! (should be easy) - all other RabbitMQMailQueue test passes

 - I believe we need a test about mail queue starvation behavior - that will be the driving force for adotion...

 - Polish?

### CODE REVIEW

**Comment by chibenwa** [comment]
@Arsnael (on monday) can you confirm we do not get starvation on top of this branch?

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
`No implementation for reactor.core.publisher.Mono<com.rabbitmq.client.Connection> was bound.`

**Comment by chibenwa** [comment]
```
[bde167bca444adae512ce66802205d8f3bbe9e36] [ERROR] Failures: 
[bde167bca444adae512ce66802205d8f3bbe9e36] [ERROR]   MemoryTerminationSubscriberTest.multipleListeningEventsShouldShareEvents 
[bde167bca444adae512ce66802205d8f3bbe9e36] Expecting:
[bde167bca444adae512ce66802205d8f3bbe9e36]   <[Failed(TaskAggregateId(TaskId{value=23d5d511-f776-428a-9109-f3eae491f417}),EventId{value=42}),
[bde167bca444adae512ce66802205d8f3bbe9e36]     Cancelled(TaskAggregateId(TaskId{value=89fc6506-ceaf-4667-a63e-eeaab7cc93ee}),EventId{value=42})]>
[bde167bca444adae512ce66802205d8f3bbe9e36] to contain exactly (and in same order):
[bde167bca444adae512ce66802205d8f3bbe9e36]   <[Completed(TaskAggregateId(TaskId{value=c58ef499-0d0f-4178-8c9e-6d457bf93c02}),EventId{value=42},COMPLETED),
[bde167bca444adae512ce66802205d8f3bbe9e36]     Failed(TaskAggregateId(TaskId{value=23d5d511-f776-428a-9109-f3eae491f417}),EventId{value=42}),
[bde167bca444adae512ce66802205d8f3bbe9e36]     Cancelled(TaskAggregateId(TaskId{value=89fc6506-ceaf-4667-a63e-eeaab7cc93ee}),EventId{value=42})]>
[bde167bca444adae512ce66802205d8f3bbe9e36] but could not find the following elements:
[bde167bca444adae512ce66802205d8f3bbe9e36]   <[Completed(TaskAggregateId(TaskId{value=c58ef499-0d0f-4178-8c9e-6d457bf93c02}),EventId{value=42},COMPLETED)]>
```

test this please

**Comment by Arsnael** [comment]
```
[bde167bca444adae512ce66802205d8f3bbe9e36] [ERROR] Failures: 
[bde167bca444adae512ce66802205d8f3bbe9e36] [ERROR]   MemoryTerminationSubscriberTest.multipleListeningEventsShouldShareEvents 
[bde167bca444adae512ce66802205d8f3bbe9e36] Expecting:
[bde167bca444adae512ce66802205d8f3bbe9e36]  <[Failed(TaskAggregateId(TaskId{value=5a9ddf15-27a7-4283-9f69-d83cd82b0412}),EventId{value=42}),
[bde167bca444adae512ce66802205d8f3bbe9e36]     Cancelled(TaskAggregateId(TaskId{value=8fbadbb3-9f17-42a5-afd3-0c098d3cfdaf}),EventId{value=42})]>
[bde167bca444adae512ce66802205d8f3bbe9e36] to be equal to:
[bde167bca444adae512ce66802205d8f3bbe9e36]  <[Completed(TaskAggregateId(TaskId{value=6af53aa4-6a9a-4354-9671-ddd2b99faa87}),EventId{value=42},COMPLETED),
[bde167bca444adae512ce66802205d8f3bbe9e36]     Failed(TaskAggregateId(TaskId{value=5a9ddf15-27a7-4283-9f69-d83cd82b0412}),EventId{value=42}),
[bde167bca444adae512ce66802205d8f3bbe9e36]     Cancelled(TaskAggregateId(TaskId{value=8fbadbb3-9f17-42a5-afd3-0c098d3cfdaf}),EventId{value=42})]>
[bde167bca444adae512ce66802205d8f3bbe9e36] but was not.
```
test this please

**Comment by chibenwa** [comment]
```
[db656ce9fc64bd8e0a9361e213c27a71c5928061] [ERROR] Failed to execute goal on project apache-james-backends-jpa: Could not resolve dependencies for project org.apache.james:apache-james-backends-jpa:jar:3.5.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus:codehaus-parent:pom:4 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/codehaus-parent/4/codehaus-parent-4.pom.part (No such file or directory) -> [Help 1]
```

test this please

**Comment by Arsnael** [comment]
```
[da0708c4bef243fcfc83f0dc97d3089f18760bf3] [ERROR] Crashed tests:
[da0708c4bef243fcfc83f0dc97d3089f18760bf3] [ERROR] org.apache.james.mailbox.events.RabbitMQEventBusTest$ConcurrentTest
[da0708c4bef243fcfc83f0dc97d3089f18760bf3] [ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
```
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
[ERROR]   RabbitMQEventBusTest$LifeCycleTest$SingleEventBus$DispatchingWhenNetWorkIssue.dispatchShouldWorkAfterNetworkIssuesForOldRegistration:358
```

**Comment by chibenwa** [comment]
test this please

ci reboot

**Comment by Arsnael** [comment]
CI restart...

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
Fixed and rebased !

**Comment by Arsnael** [comment]
When starting james on charge-01:
```
Exception in thread "main" com.google.inject.CreationException: Unable to create injector, see the following errors:

1) Could not find a suitable constructor in org.apache.james.task.eventsourcing.distributed.RabbitMQWorkQueueSupplier. Classes must have either one (and only one) constructor annotated with @Inject or a zero-argument constructor that is not private.
  at org.apache.james.task.eventsourcing.distributed.RabbitMQWorkQueueSupplier.class(RabbitMQWorkQueueSupplier.scala:32)
  at org.apache.james.modules.DistributedTaskManagerModule.configure(DistributedTaskManagerModule.java:49) (via modules: com.google.inject.util.Modules$CombinedModule
-> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> org.apache.james.modules.DistributedTaskManagerModule)

1 error
        at com.google.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:543)
        at com.google.inject.internal.InternalInjectorCreator.initializeStatically(InternalInjectorCreator.java:159)
        at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:106)
        at com.google.inject.Guice.createInjector(Guice.java:87)
        at com.google.inject.Guice.createInjector(Guice.java:69)
        at com.google.inject.Guice.createInjector(Guice.java:59)
        at org.apache.james.GuiceJamesServer.start(GuiceJamesServer.java:79)
        at org.apache.james.CassandraRabbitMQJamesServerMain.main(CassandraRabbitMQJamesServerMain.java:52)
```
Note: was the same before I fix something else and rebase. Will investigate...

File: server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java [hunk 1]
```diff
@@ -91,6 +91,7 @@ public void start() {
     }
 
     private void startWorkqueue() {
+        channelPool.start();
```
Comment (Arsnael) [inline server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java]: I added that cause with the distributed task manager bindings, the sender was null in this case... But i'm not sure it's the right way to do

Comment (chibenwa) [reply server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java]: You are rigth, that is the wrong way to fix this.

This is definitly a `Startable` intialization order. Because RabbitMQWorkQueue is not called by a Guice InitalizationOperation, then Guice have no clue that it should be started after the RabbitMQChannelPool.

The right way of doing this is to no longer call RabbitMQWorkQueue::start in RabbitMQWorkQueueSupplier

Instead, we should call it manually in tests, and rely on an initializationOperation within DistributedTaskManagerModule.

Comment (Arsnael) [reply server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java]: Yes I tried a bit this way yesterday but I lacked time, I just wanted something quick to test... Probably shouldn't have !

**Comment by chibenwa** [comment]
```
[57937f596bc7ea0125f853c88ca37ce8910000a8] [ERROR] Failures: 
[57937f596bc7ea0125f853c88ca37ce8910000a8] [ERROR]   RabbitMQEventBusTest$LifeCycleTest$SingleEventBus$DispatchingWhenNetWorkIssue.dispatchShouldWorkAfterNetworkIssuesForOldRegistration:358 
[57937f596bc7ea0125f853c88ca37ce8910000a8] Expecting code to raise a throwable.
```

**Comment by chibenwa** [comment]
Woot :+1:

File: server/container/guice/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/JMAPCommonModule.java [hunk 2]
```diff
@@ -96,4 +99,11 @@ protected void configure() {
                 accessTokenAuthenticationStrategy,
                 queryParameterAuthenticationStrategy);
     }
+
+    @ProvidesIntoSet
+    InitializationOperation workQueue(MailSpool instance) {
```
Comment (rouazana) [inline server/container/guice/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/JMAPCommonModule.java]: why is this in jmap? It does not seem related...

File: server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java [hunk 3]
```diff
@@ -203,12 +196,9 @@ public void cancel(TaskId taskId) {
     public void close() {
         Optional.ofNullable(receiverHandle).ifPresent(Disposable::dispose);
         Optional.ofNullable(receiver).ifPresent(RabbitMQExclusiveConsumer::close);
-        Optional.ofNullable(sender).ifPresent(Sender::close);
         Optional.ofNullable(sendCancelRequestsQueueHandle).ifPresent(Disposable::dispose);
         Optional.ofNullable(cancelRequestListenerHandle).ifPresent(Disposable::dispose);
-        Optional.ofNullable(sender).ifPresent(Sender::close);
         Optional.ofNullable(cancelRequestSender).ifPresent(Sender::close);
         Optional.ofNullable(cancelRequestListener).ifPresent(Receiver::close);
-        channelPool.close();
```
Comment (rouazana) [inline server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java]: why don't you need anymore to close these resources?

Comment (chibenwa) [reply server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java]: Will be closed by guice

Comment (rouazana) [reply server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java]: how is it called it tests?

Comment (chibenwa) [reply server/task/task-distributed/src/main/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueue.java]: The channelPool is no longer initialized within this class but directly in the testing @BeforeEach code hence it's no longer the responsability of this code to close it but of the testing code in a @AfterEach.

File: server/task/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueueTest.java [hunk 4]
```diff
@@ -101,13 +103,16 @@ public void close() throws IOException {
...
+        reactorRabbitMQChannelPool = new ReactorRabbitMQChannelPool(rabbitMQExtension.getRabbitConnectionPool());
+        reactorRabbitMQChannelPool.start();
+        testee = new RabbitMQWorkQueue(worker, reactorRabbitMQChannelPool, serializer);
         testee.start();
     }
 
     @AfterEach
     void tearDown() {
         testee.close();
+        reactorRabbitMQChannelPool.close();
```
Comment (chibenwa) [inline server/task/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/RabbitMQWorkQueueTest.java]: It is closed here

File: server/task/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java [hunk 5]
```diff
@@ -147,6 +152,7 @@ void setUp(EventStore eventStore) {
     void tearDown() {
         terminationSubscribers.forEach(RabbitMQTerminationSubscriber::close);
         workQueueSupplier.stopWorkQueues();
+        reactorRabbitMQChannelPool.close();
```
Comment (chibenwa) [inline server/task/task-distributed/src/test/java/org/apache/james/task/eventsourcing/distributed/DistributedTaskManagerTest.java]: And here

File: server/container/guice/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/JMAPCommonModule.java (hunk 2 above)
Comment (chibenwa) [reply server/container/guice/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/JMAPCommonModule.java]: The sin of this class (MailSpool) is generating the mailQueue within the constructor.

The asumption of some code is that I can call the MailQueueFactory within it, which means I don't need to init the MailQueue.

Brittle assumption that I break (as MQ depends on the rabbitChannel pool that should be initialized).

To be sum up, I find it pretty nice that we use the init stage of our app to create connections...

**Comment by Arsnael** [comment]
On `james-charge-01`, provisioning with maybe around 13000 mails (mini corpus, so we don't spend 3 days running that), with the RandomStoring mailet activated : 

* master branch : 
master-grafana

I did two runs but they are similar. RabbitMQ enqueues mails until it starts dequeuing and freezes. I don't know if it's the RandomStoring that is too stressful, but that's what happens in this case.

* reactive RabbitMQ (this PR) : 
rabbit-grafana

That's the result of the second run. No freeze this time, we seem to able to deliver everything in this corner case.

**Comment by Arsnael** [comment]
was :green_apple: before CI restart...

**Comment by chibenwa** [comment]
Latest commits are refactoring, mostly...

**Comment by chibenwa** [comment]
test this please

(build output lost)

**Comment by Arsnael** [comment]
CI restart...

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged
